blog/_src/posts/2019-08-19-scoping-in-r.md

```diff
@@ -0,0 +1,491 @@
+
+#### Simulating dynamic scope
+
+```r
+x <- 1
+f <- function() {
+  get("x", envir = parent.frame())
+}
+g <- function() {
+  x <- 2
+  f()
+}
+g() # 2
+```
+
+On line 3, we perform an explicit variable lookup for `x`, but we do so in the
+environment `parent.frame()`, which refers to the calling function's
+environment, in this case, `g`'s environment.. Therefore, the lookup returns
+`2`.
+
+Note that R has a similarly named function, `parent.env(e)` which returns the
+_enclosing environment_ of the given environment `e`.
+
+#### Constructing an arbitrary environment
+
+```r
+x <- 1
+f <- function() {
+  e <- new.env()
+  e$x <- 3
+  get("x", envir = e)
+}
+g <- function() {
+  x <- 2
+  f()
+}
+g() # 3
+```
+
+When `f` is called, it constructs a new environment, `e`, which is initially
+empty. (By default, its enclosing environment is the current environment, which
+is `f`'s.) Next, on line 4, it directly adds a mapping to that environment,
+assigning `3` to `x`. Then, on line 5, the lookup is explicitly done in
+environment `e`, so `f` returns `3`.
+
+#### Deleting mappings
+
+```r
+x <- 1
+f <- function() {
+  rm("x", envir = parent.env(environment()))
+  x
+}
+g <- function() {
+  x <- 2
+  f()
+}
+g() # Error in f() : object 'x' not found
+```
+
+Not only is it possible to dynamically add and modify mappings in R, but it is
+also possible to _delete_ mappings. This is what line 3 does: it explicitly
+removes the mapping for `x` from the enclosing environment of the current
+environment. In other words, the definition on line 1 is deleted. Therefore,
+when `f` is called, the lookup of `x` fails and an error is raised.
+
+#### Infinite loop during variable lookup
+
+```r
+enva <- new.env()
+envb <- new.env()
+parent.env(enva) <- envb
+parent.env(envb) <- enva
+f <- function() x
+environment(f) <- enva
+f()
+```
+
+In this final example, manipulation of environments allows us to create a
+function where variable lookup results in an infinite loop.
+
+On lines 1 and 2, we create new, empty environments. Both have the same
+enclosing environment, which is the top-level environment. However, on lines 3
+and 4, we modify their enclosing environments to create a cycle: `enva`'s
+enclosing environment is `envb`, and `envb`'s enclosing environment is `enva`.
+
+On line 5, we define a function with a free variable, `x`, but on line 6, we set
+`f`'s environment to be `enva`. Finally, we call `f`.
+
+When the body of `f` is evaluated, it needs to look up `x`. Lookup starts in
+`f`'s environment, which we set to be `enva`. Since no mapping for `x` is
+found, lookup continues in `enva`'s enclosing environment, which is `envb`.
+However, `envb` is also empty, so lookup continues in its enclosing
+environment, which is `enva`, and now lookup results in an infinite loop.
+
+
+### An intuition for scoping in R
+
+Some of the above examples appear to demonstrate dynamic scoping. Recall two of
```

Comment:
appear to? hmmmm

Reply:
Are you suggesting that I should strike out "appear to" and just say that the examples demonstrate dynamic scoping?

Again, I'm uncomfortable with that wording, even though it satisfies the definition I wrote in the previous post. That would mean R is both lexically and dynamically scoped, which is both confusing and wrong.

Reply:
It wasn't a suggestion, only a reaction.